### PR TITLE
Support cgroup v2

### LINKED ISSUE #4861: cgroups v2 silently breaks kubevirt
**Is this a BUG REPORT or FEATURE REQUEST?**:

> Uncomment only one, leave it on its own line:
>
/kind bug
> /kind enhancement

**What happened**:
VMIs get stuck on "Scheduled", and the pods backing the VMI keep restarting, with no error message.
The virt-handler errors out with "Could not detect slice of whitelisted controller: [devices]"

**What you expected to happen**:
VMs just work.

**How to reproduce it (as minimally and precisely as possible)**:
Run kubevirt using cri-o on systemd with unified cgroup hierarchy enabled (as is default in multiple distros), and try to run any VM.

**Anything else we need to know?**:

**Environment**:
- KubeVirt version (use `virtctl version`): v0.37.0
- Kubernetes version (use `kubectl version`): v1.19.7
- VM or VMI specifications: Any.
- Cloud provider or hardware configuration: almost-plain qemu on AMD. 
- OS (e.g. from /etc/os-release): NixOS 21.03pre-git.
- Kernel (e.g. `uname -a`): 5.4.86
- Install tools: kubeadm, and manual kubectl apply
- Others:

### PR DESCRIPTION
**Pull request by vasiliy-ul**
**What this PR does / why we need it**:

The PR is an **initial** attempt to enable running of Kubevirt on the systems with cgroup v2 enabled.

https://www.kernel.org/doc/html/latest/admin-guide/cgroup-v2.html

The new cgroup v2 interface is incompatible with v1 and requires additional handling. The following issues are addressed within this PR:

- cgroup slice detection in `isolation` module: the logic is based on parsing `/proc/<pid>/cgroup` file. In case of enabled v2 there is a special entry that should be handled: `0::$PATH` (this affects `ContainerDisks` and `HotplugDisks`)
- reading CPU set and other cgroup controller files: unlike v1, cgroup v2 uses a single filesystem hierarchy therefore the paths are different (affects `vCPU pinning`, several tests)

Apart from that there is a known problem with cgroup device controller and it **will be addressed separately**:
- whitelisting block devices: cgroup v2 drops device cotntroller in favor of eBPF. New implementation is needed to attach/detach eBPF programs (affects `HotplugDisks`)

Also the PR does not cover the case when both v1 and v2 are enabled. Even though the kernel allows such configuration, its not fully supported by container runtimes. So for the moment its out of the scope.

For e2e testing the corresponding cluster configuration as well as additional jobs (presubmit and periodic) have been provided:

- https://github.com/kubevirt/kubevirtci/pull/567 (configure cgroupv2 on CI nodes with k8s 1.20 cluster) 
- https://github.com/kubevirt/project-infra/pull/1092 (enable presubmit job which can be triggered manually on PRs)
- https://github.com/kubevirt/project-infra/pull/1100 (enable peridic job to keep track of the status)

The CI jobs are configured with `KUBEVIRT_E2E_SKIP='Multus|SRIOV|GPU|Macvtap|Operator'` threfore not all the tests are currently executed: 156 skipped, 586 passing (total: 742).

Fixes #4861
Follow-up issue: https://github.com/kubevirt/kubevirt/issues/5336

**Special notes for your reviewer**:

**Release note**:

```release-note
Initial cgroupv2 support
```

### CODE REVIEW

**Comment by rmohr** [comment]
> This is not fully tested yet. Enabling the tests is still in TODO. Current CI configuration is based on cgroup v1. In order to enable v2 there is a need to prepare a new VM with recent kernel and conainer runtime for executing the functests. _Ideally_ all the functest should be run against both v1 and v2 configurations. But that will double the load on CI. Probably we can start with some limited testing. Any thoughts here?

We are switching the first lanes to crio: https://github.com/kubevirt/kubevirtci/pull/524. Maybe that helps.

We could at least start having a periodic which chesk v2 for now, and when v1 is dominant maybe switching it.

File: pkg/util/cgroup/cgroup.go
Comment (rmohr) [inline pkg/util/cgroup/cgroup.go:76]: Maybe we should spilt the controllers lookup from the slice detection somehow? it looks like this circumvents some of the cgroupv1 checks. The logic is too hard for me to follow now.

File: pkg/util/cgroup/cgroup.go
Comment (rmohr) [inline pkg/util/cgroup/cgroup.go:61]: Maybe, like for the isolation detection, we can mock this file for unit testing. Now with cgroupv2, doing changes here becomes very delicate.

**Review by rmohr — CHANGES_REQUESTED** [review]
Very nice. Some intial thoughts.

File: pkg/util/cgroup/cgroup.go
Comment (vasiliy-ul) [reply pkg/util/cgroup/cgroup.go:1]: Yeah, probably you are right. The functionality is likely specific to virt-handler. Will move it, no objections.

File: pkg/util/cgroup/cgroup.go
Comment (vasiliy-ul) [reply pkg/util/cgroup/cgroup.go:76]: Agree. There is a nice set of helpers In `runc/libcontainer/cgroups` and I was thinking about refactoring the implementation using those.

File: pkg/util/cgroup/cgroup.go
Comment (vasiliy-ul) [reply pkg/util/cgroup/cgroup.go:61]: Sure, will see how to make the funcion more unittest-friendly.

**Comment by vasiliy-ul** [comment]
> 
> We are switching the first lanes to crio: [kubevirt/kubevirtci#524](https://github.com/kubevirt/kubevirtci/pull/524). Maybe that helps.
> 

I still need to setup a test VM with cgroup-v2. Just to verify it locally. Then probably it will be more clear about the dependencies. Things to consider (https://github.com/opencontainers/runc/blob/master/docs/cgroup-v2.md):
- If not enabled by default, cgroups v2 can be forced with the kernel cmdline arg: `systemd.unified_cgroup_hierarchy=1`
- kernel >= 4.15 (better >= 5.2); systemd >= 244

> We could at least start having a periodic which chesk v2 for now, and when v1 is dominant maybe switching it.

Yeah, probably better to start with some limited testing. Cgroup-v2 does not seem to be widely adopted atm.

File: pkg/util/cgroup/cgroup.go
Comment (vasiliy-ul) [reply pkg/util/cgroup/cgroup.go:1]: On the other hand, `cgroup` is also used in virt-launcher (`GetPodCPUSet` function). That in turn makes it more generic I guess. Does it still make sense to move it then? I am not opposed to that. I think that by default I followed the principle that the code used by several modules should go rather to common `util` package. But probably its not always the case.

Comment (rmohr) [reply pkg/util/cgroup/cgroup.go:1]: fine by me. I just wonder if `GetPodCPUSet` has a future there on the long run when we continue removing privileges.

Comment (vasiliy-ul) [reply pkg/util/cgroup/cgroup.go:1]: Hm... if looking from that perspective then I agree, it makes sense to move it.

Comment (rmohr) [reply pkg/util/cgroup/cgroup.go:1]: @vladikr FYI

**Comment by vasiliy-ul** [comment]
Converted the PR to draft as it seems there are more issues with cgroup v2. Investigating those. Now there are problems with accessing the devices `/dev/kvm` and `/dev/vhost-net` from `libvirt`. 

Still in TODO:
- unit-tests for `cgroup`

File: pkg/virt-handler/hotplug-disk/mount_test.go
Comment (vasiliy-ul) [inline pkg/virt-handler/hotplug-disk/mount_test.go:207]: missed that: need to remove the commented code.

**Comment by vasiliy-ul** [comment]
I think now the PR should be ready for review. Addressed the comments and also addted tests.

File: pkg/virt-handler/device-manager/generic_device.go
Comment (rmohr) [inline pkg/virt-handler/device-manager/generic_device.go:263]: I see that we add `mknod` ability now. Do we need it now for some reason?

File: pkg/virt-handler/cgroup/cgroup.go
Comment (rmohr) [inline pkg/virt-handler/cgroup/cgroup.go:231]: would you mind making `loadAttachCgroupDeviceFilter` a parameter of this function and define the function in the `v2DeviceController` struct? Or maybe making `attachDeviceFilter` a method of `v2DeviceController`?

Constructing clear objects for tests makes tests more maintainable compared to exchanging functions on well known places.

File: pkg/virt-handler/cgroup/cgroup.go
Comment (rmohr) [inline pkg/virt-handler/cgroup/cgroup.go:62]: Like in my other comment regarding to `loadAttachCgroupDeviceFilter` it would be great if we can have a clear side-effect free solution for testing.

**Review by rmohr — CHANGES_REQUESTED** [review]
Excellent PR! Two minor requests and one question inline.

File: pkg/virt-handler/device-manager/generic_device.go
Comment (rmohr) [reply pkg/virt-handler/device-manager/generic_device.go:263]: Ah found the explanation. Should we distinguish between `v1` and `v2` here and in one case add the `m` and in the other case not?

Comment (vasiliy-ul) [reply pkg/virt-handler/device-manager/generic_device.go:263]: I was thinking about it as well. I wanted to avoid having smth like `if v1 {...} else {...}` accross the code. Maybe I could introduce one more function in `cgroup` package which would return the needed permissions for the device... but imho that would look a bit weird. If distinguishing here between `v1` and `v2` I would probably go with the `if` solution as the most straightforward one. Plus some comment in the code why it is needed. At least it will be clear that its a workaround.

And In fact I've created a PR in upstream `runc` to fix the ebpf program. The PR is already merged. With the fix applied the `mknod` permission is not required and the behaviour is the same as with `v1` (so this change becomes redundant). My assumption so far has been that compatibility with older `runc` versions should be preserved. Is it the case here?

File: pkg/virt-handler/cgroup/cgroup.go
Comment (vasiliy-ul) [reply pkg/virt-handler/cgroup/cgroup.go]: Agree. I've refactored the code and created a separate commit for easier review. If its fine can leave it like that or squash afterward.

File: pkg/virt-handler/cgroup/cgroup_test.go
Comment (rmohr) [inline pkg/virt-handler/cgroup/cgroup_test.go:78]: Do you think you can introduce structs which you can replace? I would really prefer side-effect tests where one does not have to replace package-level functions.

File: pkg/virt-handler/device-manager/generic_device.go
Comment (rmohr) [inline pkg/virt-handler/device-manager/generic_device.go:265]: I would expect to make this configurable on the GenericDevicePlugin for better testability.

**Review by rmohr — COMMENTED** [review]
Looks pretty much good to me for a first pass on cgroup2 work. @davidvossel would you mind skimming through this PR too?

File: pkg/virt-handler/cgroup/cgroup_test.go
Comment (vasiliy-ul) [reply pkg/virt-handler/cgroup/cgroup_test.go:78]: hm.. not sure if I understood correctly. `cgroupFS` is a tests-local var defined above. It is then passed to the constructor of the respective struct. Currently there is no function replacement. Or maybe I missed something?

Comment (rmohr) [reply pkg/virt-handler/cgroup/cgroup_test.go:78]: Ah yes. Perfect. Never mind then :)

File: pkg/virt-handler/device-manager/generic_device.go
Comment (vasiliy-ul) [reply pkg/virt-handler/device-manager/generic_device.go:265]: I think I can change the implementation and pass the permissions via `NewGenericDevicePlugin`. But then probably it makes sense to propagate them to `NewDeviceController` from `NewController` in `vm.go`. So it will be two levels above in the callstack. What do you think? Is it okay to do like that?

**Comment by davidvossel** [comment]
> Looks pretty much good to me for a first pass on cgroup2 work. @davidvossel would you mind skimming through this PR too?

@rmohr I'd like to see a CI lane with functionality passing before i get too involved. Nothing jumped out at me skimming through this, but i think there could be more work to do.

@awels fyi, your block pvc hotplug will likely impacted by this.

> But that will double the load on CI. Probably we can start with some limited testing. Any thoughts here?

@vasiliy-ul I think initially it's enough to have one lane that executes using cgroup v2 for now. We'll need that before we can merge anything

**Comment by vasiliy-ul** [comment]
> @vasiliy-ul I think initially it's enough to have one lane that executes using cgroup v2 for now. We'll need that before we can merge anything

@davidvossel , sounds reasonable. I can take a look on how to add the lane with cgroup v2.

**Comment by awels** [comment]
If you can make sure to include the rook-ceph context in tests/storage/hotplug.go in that test lane, that will capture failures to hotplug because of changes here.

**Comment by vasiliy-ul** [comment]
As I see the PR https://github.com/kubevirt/kubevirtci/pull/541 which enables crio instead of docker is already approved and hopefully will be merged soon. I think indeed it will help a lot with testing. When its merged the only thing left will be enabling cgroupv2 via the kernel cmdline `systemd.unified_cgroup_hierarchy=1`.

Currently the kernel `4.18` is used in CI (from CentOS 8) and looks like it supports cgroupv2. I looked at how to pass additional arguments to the kernel cmdline. Probably it can be done using `gocli` tool. It supports additional qemu arguments https://github.com/kubevirt/kubevirtci/blob/a2b389fadcc6dee26acc8f36c8be3d037e235f96/cluster-provision/gocli/cmd/run.go#L70 and I guess its possible to pass smth like `-append "systemd.unified_cgroup_hierarchy=1"`. I also found that there is `KUBEVIRT_PROVIDER_EXTRA_ARGS` env var that can be used.

Might it be an option then to crete a dedicated presubmit job that will add `-append ...` to qemu, run k8s 1.20 with crio and also include `rook-ceph` test context? Though dont know if I missed smth or if there is a better way... any thoughts?

**Comment by vasiliy-ul** [comment]
I think it is better to split this PR. While looking at how to enable the functional testing in CI I found a couple of issues with 'injecting' devices into an existing pod. It seems that with eBPF filtering it is now more complicated comparing to using device cgroup v1 controller. So I plan to address that separately otherwise this PR tends to grow. Here I want to leave only the essential fixes which are needed  for basic functionality (i.e. just to start a VM). But that means hotplug volumes will not work for now. Any objections on that @rmohr , @davidvossel , @awels ?

**Comment by awels** [comment]
Well we worked quite long and hard to enable the hotplugging feature, how long will it not work anymore if we merge the basics?

**Comment by vasiliy-ul** [comment]
> Well we worked quite long and hard to enable the hotplugging feature, how long will it not work anymore if we merge the basics?

@awels , no I mean it will not work only with cgroup **v2** (it actually does not work with such configuration even now). Current implementation works fine with cgroup **v1** and it will not be broken.

**Comment by awels** [comment]
Gotcha, no problem on my end then.

**Comment by vasiliy-ul** [comment]
For now I removed the part with eBPF filtering. As mentioned earlier this needs to be addressed separately. Here I left only the fixes to enable the basic functionality.

**Comment by rmohr** [comment]
@vasiliy-ul I think the final task is now to create an optional presubmit in https://github.com/kubevirt/project-infra/blob/34905ccc4fa3a8cd2364a4436e475a108e502860/github/ci/prow-deploy/files/jobs/kubevirt/kubevirt/kubevirt-presubmits.yaml. Once this is done we can judge the state of the PR, and maybe define a passing test-subset for a periodic job.

**Comment by vasiliy-ul** [comment]
> @vasiliy-ul I think the final task is now to create an optional presubmit in https://github.com/kubevirt/project-infra/blob/34905ccc4fa3a8cd2364a4436e475a108e502860/github/ci/prow-deploy/files/jobs/kubevirt/kubevirt/kubevirt-presubmits.yaml. Once this is done we can judge the state of the PR, and maybe define a passing test-subset for a periodic job.

@rmohr , the presubmit job is already created in https://github.com/kubevirt/project-infra/pull/1092

**Comment by rmohr** [comment]
> @rmohr , the presubmit job is already created in [kubevirt/project-infra#1092](https://github.com/kubevirt/project-infra/pull/1092)

great! Merged.

/test pull-kubevirt-e2e-k8s-1.20-cgroupsv2

**Comment by vasiliy-ul** [comment]
I think there is an error:
```
./cluster-up/up.sh
242
08:54:29: Download the image quay.io/kubevirtci/k8s-1.20-cgroupsv2:2103250939-432da7e
243
08:54:29: time="2021-03-26T08:54:29Z" level=info msg="Using remote image quay.io/kubevirtci/k8s-1.20-cgroupsv2:2103250939-432da7e"
244
08:54:30: 2021/03/26 08:54:30 failed to download quay.io/kubevirtci/k8s-1.20-cgroupsv2:2103250939-432da7e: Error response from daemon: unauthorized: access to the requested resource is not authorized
245
08:54:32: 2021/03/26 08:54:32 failed to download quay.io/kubevirtci/k8s-1.20-cgroupsv2:2103250939-432da7e: Error response from daemon: unauthorized: access to the requested resource is not authorized
246
08:54:34: 2021/03/26 08:54:34 failed to download quay.io/kubevirtci/k8s-1.20-cgroupsv2:2103250939-432da7e: Error response from daemon: unauthorized: access to the requested resource is not authorized
247
08:54:41: 2021/03/26 08:54:41 failed to download quay.io/kubevirtci/k8s-1.20-cgroupsv2:2103250939-432da7e: Error response from daemon: unauthorized: access to the requested resource is not authorized
248
08:54:41: panic: failed to download quay.io/kubevirtci/k8s-1.20-cgroupsv2:2103250939-432da7e four times, giving up.
249
08:54:41: 
```
Are we missing smth?

**Comment by rmohr** [comment]
> Are we missing smth? Maybe in the job?

It was the first push to this new repo. Seems like when quay creates a new repo implicitly via a push it is private by default. Made it public. At least that should be solved now.

/test pull-kubevirt-e2e-k8s-1.20-cgroupsv2

**Comment by vasiliy-ul** [comment]
For now only one failed tests in https://prow.apps.ovirt.org/view/gcs/kubevirt-prow/pr-logs/pull/kubevirt_kubevirt/4907/pull-kubevirt-e2e-k8s-1.20-cgroupsv2/1375376301224890368

Looks like I missed the hardcoded path which does not exist with cgroup v2: https://github.com/kubevirt/kubevirt/blob/a2f648a7c00700434101a5f944305466ede56495/tests/vmi_configuration_test.go#L626

IMHO this PR is already big enough. I would prefer to create a new PR with the fix if no objections.

Also now the job is configured with `KUBEVIRT_E2E_SKIP='Multus|SRIOV|GPU|Macvtap|Operator'` and many tests are skipped. But the basics seem to work: >300 tests passed.

**Comment by rmohr** [comment]
> Also now the job is configured with `KUBEVIRT_E2E_SKIP='Multus|SRIOV|GPU|Macvtap|Operator'` and many tests are skipped. But the basics seem to work: >300 tests passed.

That sounds like a good start to me. Let's fix the one test and finish the code review.

**Comment by rmohr** [comment]
> Also now the job is configured with `KUBEVIRT_E2E_SKIP='Multus|SRIOV|GPU|Macvtap|Operator'` and many tests are skipped. But the basics seem to work: >300 tests passed.

Oh btw. the one test failed on the parallel tests. If they fail the serial tests are not executed.  There may fail more once you fix this one.

**Comment by rmohr** [comment]
> IMHO this PR is already big enough. I would prefer to create a new PR with the fix if no objections.

I should have read all of your message. OK for me.

**Comment by vasiliy-ul** [comment]
Even better now: https://prow.apps.ovirt.org/view/gcs/kubevirt-prow/pr-logs/pull/kubevirt_kubevirt/4907/pull-kubevirt-e2e-k8s-1.20-cgroupsv2/1375435441154560000/

156 skipped, 586 passed and no failures

**Comment by rmohr** [comment]
And here another passing one after the rebase: https://prow.apps.ovirt.org/view/gcs/kubevirt-prow/pr-logs/pull/kubevirt_kubevirt/4907/pull-kubevirt-e2e-k8s-1.20-cgroupsv2/1375731264539070464

@vasiliy-ul I think you can change the cgroupv2 test lane to report its status on github (but not always_run yet) and add a periodic job which will execute cgroupv2 test daily. We can then continue with merging this PR.

**Review by rmohr — CHANGES_REQUESTED** [review]
One mini request.

File: pkg/virt-handler/cgroup/cgroup_suite_test.go
Comment (vasiliy-ul) [reply pkg/virt-handler/cgroup/cgroup_suite_test.go]: I somehow missed that. Fixed.

**Comment by rmohr** [comment]
/hold

@vasiliy-ul can you update the PR description? I think a few points should be addressed:

 * How much is tested now
 * #4861 is not a tracker issue for cgroupsv2 support, so it may be ok to close it with this PR, would you mind creating a tracker bug?
 * Can you change the release note to something like "First pass on working towards supporting cgroupsv2", to make clear that this does not yet mean full cgroupsv2 support.
 * Clearly state in the general PR description that not all work is done yet.
 
 Once, done feel free to unhold the PR :)

**Comment by vasiliy-ul** [comment]
> /hold
> 
> @vasiliy-ul can you update the PR description? I think a few points should be addressed:
> 

@rmohr , sure, will update it. Just to clarify a few things:

> * #4861 is not a tracker issue for cgroupsv2 support, so it may be ok to close it with this PR, would you mind creating a tracker bug?

Do you mean creating a bug for hotplug disks support with cgroupv2 or just a general cgroupv2 placeholder issue?

> * Can you change the release note to something like "First pass on working towards supporting cgroupsv2", to make clear that this does not yet mean full cgroupsv2 support.

I was thinking about removing the release note actually. Can be added when all the features are working. What do you think?

> * Clearly state in the general PR description that not all work is done yet.
> 

Ack.

**Comment by rmohr** [comment]
> Do you mean creating a bug for hotplug disks support with cgroupv2 or just a general cgroupv2 placeholder issue?
> 
> >

Yes, just a tracker issue, where we put everything until we say: kubevirt now supports cgroupv2 (mostly means also testing right now skipped tests and hotplut).

> I was thinking about removing the release note actually. Can be added when all the features are working. What do you think?
> 
> >

Also fine by me. It may still be nice to see that progress is made towards that goal.

**Comment by vasiliy-ul** [comment]
@rmohr , the description is updated.

**Comment by davidvossel** [comment]
i see the ci presubmit job was created in project-infra, but i don't see the results showing up here

**Comment by vasiliy-ul** [comment]
> i see the ci presubmit job was created in project-infra, but i don't see the results showing up here

@davidvossel , the results are not shown here beause the job has `skip_report: true`. I addressed that in https://github.com/kubevirt/project-infra/pull/1100 . 

The results can be observed following the link:
https://prow.apps.ovirt.org/view/gcs/kubevirt-prow/pr-logs/pull/kubevirt_kubevirt/4907/pull-kubevirt-e2e-k8s-1.20-cgroupsv2/1375435441154560000/

**Pull request merged**
